lightning/src/util/persist.rs

```diff
@@ -194,3 +206,857 @@ where
 	}
 	Ok(res)
 }
+enum KVStoreUpdatingPersisterError<'a> {
+	/// The monitor name was improperly formatted.
+	BadMonitorName { reason: &'a str, context: &'a str },
+	/// The update name was improperly formatted.
+	BadUpdateName { reason: &'a str, context: &'a str },
+	/// The monitor could not be decoded.
+	MonitorDecodeFailed {
+		reason: DecodeError,
+		context: &'a str,
+	},
+	/// The update could not be decoded.
+	UpdateDecodeFailed {
+		reason: DecodeError,
+		context: &'a str,
+	},
+	/// Storage could not be read.
+	StorageReadFailed { reason: io::Error, context: &'a str },
+	/// An update could not be applied to a monitor.
+	UpdateFailed { reason: &'a str, context: &'a str },
+}
+
+impl<'a> From<KVStoreUpdatingPersisterError<'a>> for io::Error {
+	fn from(value: KVStoreUpdatingPersisterError) -> Self {
+		match value {
+			KVStoreUpdatingPersisterError::BadMonitorName { reason, context } => io::Error::new(
+				io::ErrorKind::InvalidInput,
+				format!("BadMonitorName, {}, context: {}", reason, context),
+			),
+			KVStoreUpdatingPersisterError::BadUpdateName { reason, context } => io::Error::new(
+				io::ErrorKind::InvalidInput,
+				format!("BadUpdateName, {}, context: {}", reason, context),
+			),
+			KVStoreUpdatingPersisterError::MonitorDecodeFailed { reason, context } => {
+				io::Error::new(
+					io::ErrorKind::InvalidData,
+					format!("MonitorDecodeFailed, {}, context: {}", reason, context),
+				)
+			}
+			KVStoreUpdatingPersisterError::UpdateDecodeFailed { reason, context } => {
+				io::Error::new(
+					io::ErrorKind::InvalidData,
+					format!("UpdateDecodeFailed, {}, context: {}", reason, context),
+				)
+			}
+			KVStoreUpdatingPersisterError::StorageReadFailed { reason, context } => io::Error::new(
+				io::ErrorKind::Other,
+				format!("StorageReadFailed, {}, context: {}", reason, context),
+			),
+			KVStoreUpdatingPersisterError::UpdateFailed { reason, context } => io::Error::new(
+				io::ErrorKind::InvalidData,
+				format!("UpdateFailed, {}, context: {}", reason, context),
+			),
+		}
+	}
+}
+
+struct KVStoreUpdatingPersister<K: Deref, L: Deref>
```

Comment:
Ack on the docs. I wrote a lot for my internal implementation of this, and I planned to bring them over and tweak them once this got closer to final form. I'm looking forward to the docs pass.

> Maybe `MonitorUpdatingStore`?

Yeah the name was basically a placeholder, I am not sure what to call it. `MonitorUpdatingStor(ag)e` sounds about right. Definitely happy to give more attention to naming things, it's arguably the hardest part.

> We'll definitely also need some explanation of how the prepended sentinels work and how users might recover 'regular' monitors.

Yeah, I was thinking about this too. I was even wondering about making a utility to perform the recovery, possibly in my own GH repo since it might be difficult to include in LDK. Open to ideas about that, though.

I think the recovery-to-downgrade procedure is something like:
  - Take the node offline (you don't want to be actively writing to do this)
  - Ensure all updates have been committed to the monitors
    - Does the `BackgroundProcessor` already do this on an orderly shutdown? I think so? If so, the sentinel bytes are only serving to protect you from missing updates if you downgrade after a panic, I guess.
  - Remove the leading sentinel bytes from the monitors
  - Restart node with downgrade / standard persister